File: Android/frontend/lib/services/api_service.dart
import 'dart:convert';
import 'package:http/http.dart' as http;
import '../models/models.dart';

import 'package:flutter/foundation.dart'; // For kIsWeb

class ApiService {
  static const String baseUrl = kIsWeb ? 'http://localhost:8000' : 'http://10.197.192.17:8000'; // IP required for Emulator / Web
  static int? currentUserId = 1; // Default mock user ID since login is bypassed
  static String? currentUserRole = 'hosteler'; // Default mock role

  // --- Auth ---
  static Future<AuthResponse> login(String email, String password) async {
    final response = await http.post(
      Uri.parse('$baseUrl/login'),
      headers: {'Content-Type': 'application/json'},
      body: json.encode({'email': email, 'password': password}),
    );
    if (response.statusCode == 200) {
      final res = AuthResponse.fromJson(json.decode(response.body));
      currentUserId = res.userId;
      currentUserRole = res.role;
      return res;
    } else {
      throw Exception('Login failed: ${json.decode(response.body)['detail']}');
    }
  }

  static Future<AuthResponse> registerHosteler(Map<String, dynamic> data) async {
    final response = await http.post(
      Uri.parse('$baseUrl/register/hosteler'),
      headers: {'Content-Type': 'application/json'},
      body: json.encode(data),
    );
    if (response.statusCode == 200) {
      final res = AuthResponse.fromJson(json.decode(response.body));
      currentUserId = res.userId;
      currentUserRole = res.role;
      return res;
    } else {
      throw Exception('Registration failed: ${json.decode(response.body)['detail']}');
    }
  }

  static Future<AuthResponse> registerWarden(Map<String, dynamic> data) async {
    final response = await http.post(
      Uri.parse('$baseUrl/register/warden'),
      headers: {'Content-Type': 'application/json'},
      body: json.encode(data),
    );
    if (response.statusCode == 200) {
      final res = AuthResponse.fromJson(json.decode(response.body));
      currentUserId = res.userId;
      currentUserRole = res.role;
      return res;
    } else {
      throw Exception('Registration failed: ${json.decode(response.body)['detail']}');
    }
  }

  // --- Profile ---
  static Future<Profile> getStudentProfile() async {
    final response = await http.get(Uri.parse('$baseUrl/profile'));
    if (response.statusCode == 200) {
      return Profile.fromJson(json.decode(response.body));
    } else {
      throw Exception('Failed to load profile');
    }
  }

  static Future<Profile> updateProfile(Map<String, dynamic> data) async {
    final response = await http.put(
      Uri.parse('$baseUrl/profile/update'),
      headers: {'Content-Type': 'application/json'},
      body: json.encode(data),
    );
    if (response.statusCode == 200) {
      return Profile.fromJson(json.decode(response.body));
    } else {
      throw Exception('Failed to update profile');
    }
  }

  // --- Dashboard ---
  static Future<DashboardSummary> getDashboardSummary() async {
    final response = await http.get(Uri.parse('$baseUrl/dashboard/summary'));
    if (response.statusCode == 200) {
      return DashboardSummary.fromJson(json.decode(response.body));
    } else {
      throw Exception('Failed to load dashboard summary');
    }
  }

  // --- Gate Pass ---
  static Future<GatePass> createGatePass(Map<String, dynamic> data) async {
    final response = await http.post(
      Uri.parse('$baseUrl/gatepass/create'),
      headers: {'Content-Type': 'application/json'},
      body: json.encode(data),
    );
    if (response.statusCode == 200) {
      return GatePass.fromJson(json.decode(response.body));
    } else {
      throw Exception('Failed to create gate pass');
    }
  }

  static Future<List<GatePass>> getGatePasses() async {
    final response = await http.get(Uri.parse('$baseUrl/gatepass/list'));
    if (response.statusCode == 200) {
      Iterable l = json.decode(response.body);
      return List<GatePass>.from(l.map((model) => GatePass.fromJson(model)));
    } else {
      throw Exception('Failed to load gate passes');
    }
  }

  // --- Leave ---
  static Future<Leave> createLeave(Map<String, dynamic> data) async {
    final response = await http.post(
      Uri.parse('$baseUrl/leave/create'),
      headers: {'Content-Type': 'application/json'},
      body: json.encode(data),
    );
    if (response.statusCode == 200) {
      return Leave.fromJson(json.decode(response.body));
    } else {
      throw Exception('Failed to create leave request');
    }
  }

  static Future<List<Leave>> getLeaves() async {
    final response = await http.get(Uri.parse('$baseUrl/leave/list'));
    if (response.statusCode == 200) {
      Iterable l = json.decode(response.body);
      return List<Leave>.from(l.map((model) => Leave.fromJson(model)));
    } else {
      throw Exception('Failed to load leaves');
    }
  }

  // --- Attendance ---
  static Future<AttendanceStatus> markAttendance() async {
    final response = await http.post(Uri.parse('$baseUrl/attendance/mark'));
    if (response.statusCode == 200) {
      return AttendanceStatus.fromJson(json.decode(response.body));
    } else {
      throw Exception('Failed to mark attendance');
    }
  }

  static Future<AttendanceStatus> getAttendanceStatus() async {
    final response = await http.get(Uri.parse('$baseUrl/attendance/status'));
    if (response.statusCode == 200) {
      return AttendanceStatus.fromJson(json.decode(response.body));
    } else {
      throw Exception('Failed to load attendance status');
    }
  }

  // --- Visitor ---
  static Future<Visitor> createVisitor(Map<String, dynamic> data) async {
    final response = await http.post(
      Uri.parse('$baseUrl/visitor/create'),
      headers: {'Content-Type': 'application/json'},
      body: json.encode(data),
    );
    if (response.statusCode == 200) {
      return Visitor.fromJson(json.decode(response.body));
    } else {
      throw Exception('Failed to register visitor');
    }
  }

  static Future<List<Visitor>> getVisitors() async {
    final response = await http.get(Uri.parse('$baseUrl/visitor/list'));
    if (response.statusCode == 200) {
      Iterable l = json.decode(response.body);
      return List<Visitor>.from(l.map((model) => Visitor.fromJson(model)));
    } else {
      throw Exception('Failed to load visitors');
    }
  }

  // --- Complaint ---
  static Future<Complaint> createComplaint(Map<String, dynamic> data) async {
    final response = await http.post(
      Uri.parse('$baseUrl/complaint/create'),
      headers: {'Content-Type': 'application/json'},
      body: json.encode(data),
    );
    if (response.statusCode == 200) {
      return Complaint.fromJson(json.decode(response.body));
    } else {
      throw Exception('Failed to log complaint');
    }
  }

  static Future<List<Complaint>> getComplaints() async {
    final response = await http.get(Uri.parse('$baseUrl/complaint/list'));
    if (response.statusCode == 200) {
      Iterable l = json.decode(response.body);
      return List<Complaint>.from(l.map((model) => Complaint.fromJson(model)));
    } else {
      throw Exception('Failed to load complaints');
    }
  }

  // --- Vote ---
  static Future<VoteItem> submitVote(String menuItem) async {
    final response = await http.post(
      Uri.parse('$baseUrl/vote'),
      headers: {'Content-Type': 'application/json'},
      body: json.encode({'menu_item': menuItem}),
    );
    if (response.statusCode == 200) {
      return VoteItem.fromJson(json.decode(response.body));
    } else {
      throw Exception('Failed to submit vote');
    }
  }

  static Future<List<VoteItem>> getVoteResults() async {
    final response = await http.get(Uri.parse('$baseUrl/vote/results'));
    if (response.statusCode == 200) {
      Iterable l = json.decode(response.body);
      return List<VoteItem>.from(l.map((model) => VoteItem.fromJson(model)));
    } else {
      throw Exception('Failed to load vote results');
    }
  }

  // --- Notifications ---
  static Future<List<NotificationItem>> getNotifications() async {
    final response = await http.get(Uri.parse('$baseUrl/notifications/list'));
    if (response.statusCode == 200) {
      Iterable l = json.decode(response.body);
      return List<NotificationItem>.from(l.map((model) => NotificationItem.fromJson(model)));
    } else {
      throw Exception('Failed to load notifications');
    }
  }
}

File: Web/assets/js/api_service.js
// ========================================
// HostelSync – Central API Service
// ========================================

const API_BASE = 'http://10.197.192.17:8000';

class ApiService {
  constructor() {
    this.baseUrl = API_BASE;
  }

  getToken() {
    return localStorage.getItem('hostelsync_token');
  }

  getHeaders(withAuth = true) {
    const headers = { 'Content-Type': 'application/json' };
    if (withAuth) {
      const token = this.getToken();
      if (token) headers['Authorization'] = `Bearer ${token}`;
    }
    return headers;
  }

  async request(method, endpoint, data = null, withAuth = true) {
    const url = `${this.baseUrl}${endpoint}`;
    const options = {
      method,
      headers: this.getHeaders(withAuth),
    };
    if (data && (method === 'POST' || method === 'PUT' || method === 'PATCH')) {
      options.body = JSON.stringify(data);
    }
    try {
      const response = await fetch(url, options);
      const result = await response.json();
      if (!response.ok) {
        throw new Error(result.detail || result.message || 'Request failed');
      }
      return result;
    } catch (error) {
      if (error.message === 'Failed to fetch') {
        showToast('Server unavailable. Please check your connection.', 'error');
      } else {
        showToast(error.message, 'error');
      }
      throw error;
    }
  }

  // ===== AUTH =====
  async login(email, password) {
    const result = await this.request('POST', '/login', { email, password }, false);
    if (result.accessToken) {
      localStorage.setItem('hostelsync_token', result.accessToken);
      localStorage.setItem('hostelsync_user', JSON.stringify({
        userId: result.userId,
        role: result.role,
        name: result.name,
        hostelId: result.hostelId,
      }));
    }
    return result;
  }

  async registerWarden(data) {
    return this.request('POST', '/register/warden', data, false);
  }

  logout() {
    localStorage.removeItem('hostelsync_token');
    localStorage.removeItem('hostelsync_user');
    window.location.href = 'index.html';
  }

  getUser() {
    const user = localStorage.getItem('hostelsync_user');
    return user ? JSON.parse(user) : null;
  }

  isAuthenticated() {
    return !!this.getToken();
  }

  // ===== GATE PASS =====
  async createGatePass(data) {
    return this.request('POST', '/gatepass', data);
  }

  async getGatePassList(params = {}) {
    const query = new URLSearchParams(params).toString();
    return this.request('GET', `/gatepass${query ? '?' + query : ''}`);
  }

  async approveGatePass(id) {
    return this.request('PUT', `/gatepass/${id}/approve`);
  }

  async rejectGatePass(id) {
    return this.request('PUT', `/gatepass/${id}/reject`);
  }

  // ===== LEAVE =====
  async createLeave(data) {
    return this.request('POST', '/leave', data);
  }

  async getLeaves(params = {}) {
    const query = new URLSearchParams(params).toString();
    return this.request('GET', `/leave${query ? '?' + query : ''}`);
  }

  async approveLeave(id) {
    return this.request('PUT', `/leave/${id}/approve`);
  }

  async rejectLeave(id) {
    return this.request('PUT', `/leave/${id}/reject`);
  }

  // ===== ATTENDANCE =====
  async markAttendance(data) {
    return this.request('POST', '/attendance', data);
  }

  async getAttendanceStatus(params = {}) {
    const query = new URLSearchParams(params).toString();
    return this.request('GET', `/attendance${query ? '?' + query : ''}`);
  }

  // ===== COMPLAINTS =====
  async createComplaint(data) {
    return this.request('POST', '/complaint', data);
  }

  async getComplaints(params = {}) {
    const query = new URLSearchParams(params).toString();
    return this.request('GET', `/complaint${query ? '?' + query : ''}`);
  }

  async updateComplaintStatus(id, status) {
    return this.request('PUT', `/complaint/${id}`, { status });
  }

  // ===== VISITORS =====
  async createVisitor(data) {
    return this.request('POST', '/visitor', data);
  }

  async getVisitors(params = {}) {
    const query = new URLSearchParams(params).toString();
    return this.request('GET', `/visitor${query ? '?' + query : ''}`);
  }

  // ===== VOTING =====
  async vote(data) {
    return this.request('POST', '/vote', data);
  }

  async getResults(params = {}) {
    const query = new URLSearchParams(params).toString();
    return this.request('GET', `/vote/results${query ? '?' + query : ''}`);
  }

  // ===== NOTIFICATIONS =====
  async getNotifications() {
    return this.request('GET', '/notifications');
  }

  // ===== DASHBOARD =====
  async getDashboardStats() {
    return this.request('GET', '/dashboard/stats');
  }
}

// ===== Toast notification =====
function showToast(message, type = 'success') {
  let container = document.querySelector('.toast-container');
  if (!container) {
    container = document.createElement('div');
    container.className = 'toast-container';
    document.body.appendChild(container);
  }
  const toast = document.createElement('div');
  toast.className = `toast toast--${type}`;
  toast.textContent = message;
  container.appendChild(toast);
  setTimeout(() => {
    toast.style.opacity = '0';
    toast.style.transform = 'translateX(40px)';
    toast.style.transition = 'all 0.3s';
    setTimeout(() => toast.remove(), 300);
  }, 3500);
}

// Global instance
const api = new ApiService();
